proxy/http/HttpSM.cc

```diff
@@ -4860,56 +4868,99 @@ HttpSM::do_http_server_open(bool raw)
                                    static_cast<int>(strlen(t_state.current.server->name)));
 
-    auto ccount = connections->getCount(t_state.current.server->dst_addr, hostname_hash,
-                                        (TSServerSessionSharingMatchType)t_state.txn_conf->server_session_sharing_match);
-    if (ccount >= t_state.txn_conf->origin_max_connections) {
-      ip_port_text_buffer addrbuf;
-      ats_ip_nptop(&t_state.current.server->dst_addr.sa, addrbuf, sizeof(addrbuf));
-      SMDebug("http", "[%" PRId64 "] too many connections (%d) for this host (%" PRId64 "): %s", sm_id, ccount,
-              t_state.txn_conf->origin_max_connections, addrbuf);
-      Warning("[%" PRId64 "] too many connections (%d) for this host (%" PRId64 "): %s", sm_id, ccount,
-              t_state.txn_conf->origin_max_connections, addrbuf);
-      ink_assert(pending_action == nullptr);
+    // ugly - if match is NONE then outbound connection tracking is de facto disabled because there
+    // are no common sessions to count. Try to back up to find a more general match until we get
+    // to a hardwired value that's not NONE.
+    TSServerSessionSharingMatchType mt =
+      static_cast<TSServerSessionSharingMatchType>(t_state.txn_conf->server_session_sharing_match);
+    if (TS_SERVER_SESSION_SHARING_MATCH_NONE == mt) {
+      mt = static_cast<TSServerSessionSharingMatchType>(t_state.http_config_param->oride.server_session_sharing_match);
+      if (TS_SERVER_SESSION_SHARING_MATCH_NONE == mt) {
+        // Really? Every single session is private *and* outbound connection tracking is on?
+        mt = TS_SERVER_SESSION_SHARING_MATCH_IP; // so use this, least worst choice.
+      }
+    }
+
+    t_state.conn_tracker_info = OutboundConnTracker::get(t_state.current.server->dst_addr, hostname_hash, mt);
+  }
 
-      // if we were previously queued, or the queue is disabled-- just reschedule
-      if (t_state.origin_request_queued || t_state.txn_conf->origin_max_connections_queue < 0) {
+  // Check to see if we have reached the max number of connections on this upstream host.
+  if (t_state.txn_conf->origin_max_connections > 0) {
+    auto &ct_info = *t_state.conn_tracker_info;
+    std::time_t lat;                                   // last alert time.
+    auto ccount                    = ++ct_info._count; // cache instantaneous value plus reservation for this.
+    t_state.outbound_conn_reserved = true;
+    if (ccount > t_state.txn_conf->origin_max_connections) {
+      --ct_info._count; // cancel our reservation
+      t_state.outbound_conn_reserved = false;
+
+      ink_assert(pending_action == nullptr); // in case of reschedule must not have already pending.
+
+      // If the queue is disabled, reschedule.
+      if (t_state.txn_conf->origin_max_connections_queue < 0) {
+        ++ct_info._rescheduled;
         pending_action = eventProcessor.schedule_in(this, HRTIME_MSECONDS(100));
```

Comment:
This is a bad default and I wonder if the option should even exist.  Allowing an unlimited number state machines to be rescheduled every 100ms with no upper bounds on the number of times to be rescheduled is not a good idea.

Reply:
That's how the previous version worked, I was trying to minimize functional changes.

Reply:
Since it is rewritten, the design should be looked at instead of just copying the existing functionality.